### DISCUSSION TITLE
RelationshipId sometimes changes on refresh

### DISCUSSION THREAD

**Discussion by bleathome** [structurizr/lite#49] [Q&A] [answered]
Hello, using Structurizr Lite and working on a plug-in, I realized that RelationshipId change on refresh. Is it the desired behavior ? It is possible to guarantee a constant json when the DSL remains the same ?

**Comment by simonbrowndotje** [comment]
This is as designed ... the underlying Structurizr for Java library uses a [sequential number](https://github.com/structurizr/java/blob/master/structurizr-core/src/com/structurizr/model/SequentialIntegerIdGeneratorStrategy.java) as the basis for element/relationship IDs, so those IDs will change depending upon the order of lines in the DSL.

If you would like to change this behaviour, you can write your own implementation of [IdGenerator](https://github.com/structurizr/java/blob/master/structurizr-core/src/com/structurizr/model/IdGenerator.java) and [plug it in to the model](https://github.com/structurizr/java/blob/master/structurizr-core/src/com/structurizr/model/Model.java#L1016).

**Reply by bleathome** [reply]
Thanks. The iDs change depending upon the order of lines in the DSL. That's ok for me. But, there are case when they change with a unmodified DSL file.

**Comment by simonbrowndotje** [comment]
> But, there are case when they change with a unmodified DSL file.

I would need to check, but that may be the case with (1) implied relationships and (2) relationships between software system/container instances.

**Reply by bleathome** [reply]
Using this DSL and only refreshing the browser view, the json file sometimes changes. Is it necessary to parse again a DSL that didnt change ? (without plug-in)
[workspace.zip](https://github.com/structurizr/lite/files/11120952/workspace.zip).

**Reply by simonbrowndotje** [reply]
As I said before, if you need stable internal IDs (that don't change between DSL refreshes), I'd recommend building your own `IdGenerator`.

What exactly is your plugin trying to do, and why are you relying on internal relationship IDs?

**Reply by bleathome** [reply]
My plugin numbers the relations of each deployment view and builds a table of relations in asciidoc format. Unfortunately, in some cases when I switch between diagram and documentation views, the relationship numbers have changed and no longer match. Here is a DSL file and a simplified version of the plugin that highlights my problem. 
I don't rely on the internal relationship Ids, I just list the relationships through the API and the list is not always in the same order while the DSL has not been changed.
[Plugin.zip](https://github.com/structurizr/lite/files/11125269/Plugin.zip)

**Reply by simonbrowndotje** [reply, marked as answer]
> I would need to check, but that may be the case with (1) implied relationships and (2) relationships between software system/container instances.

This is the case ... with (2) the [replicateElementRelationships()](https://github.com/structurizr/java/blob/master/structurizr-core/src/com/structurizr/model/Model.java#L952) method finds eligible elements via a Java stream and does a `toSet()`, which won't guarantee the ordering, and therefore relationships may be allocated different IDs.

Why the change only sometimes? Changes made in the diagram editor in your web browser will cause the workspace to be saved (i.e. saving the layout information). The underlying workspace is stored as a JSON file, and the Java serialisation process (from the DSL parser) will sometimes differ from the JavaScript serialisation process (from the diagram editor), particularly with regards to ordering of objects.

I took a look at your plugin, and modifying the description of relationship views is also going to cause some issues (particularly with the layout merging algorithm that's built into the Structurizr for Java library that sits behind the DSL parser, Structurizr Lite, etc). It's an oversight that the `setDescription()` method is public, which I will fix in the next release.
